Python/Product/Wsl/Debugger/WslLauncher.cs

```diff
@@ -0,0 +1,327 @@
+            var debugger3 = (EnvDTE90.Debugger3)dte.Debugger;
+            var transports = debugger3.Transports;
+            EnvDTE80.Transport transport = null;
+            for (int i = 1; i <= transports.Count; ++i) {
+                var t = transports.Item(i);
+                if (Guid.Parse(t.ID) == portSupplier) {
+                    transport = t;
+                    break;
+                }
+            }
+            if (transport == null) {
+                return false;
+            }
+
+            var processes = debugger3.GetProcesses(transport, transportQualifierUri);
+            if (processes.Count < 1) {
+                return false;
+            }
+
+            var process = processes.Item(1);
+            return AttachToProcess(processOutput, process);
+        }
+
+        private static bool AttachToProcess(Process processOutput, EnvDTE.Process process, Guid[] engines = null) {
+            // Retry the attach itself 3 times before displaying a Retry/Cancel
+            // dialog to the user.
+            var dte = process.DTE;
+            dte.SuppressUI = true;
+            try {
+                try {
+                    if (engines == null) {
+                        process.Attach();
+                    } else {
+                        var process3 = process as EnvDTE90.Process3;
+                        if (process3 == null) {
+                            return false;
+                        }
+                        process3.Attach2(engines.Select(engine => engine.ToString("B")).ToArray());
+                    }
+                    return true;
+                } catch (COMException) {
+                    if (processOutput.WaitForExit(500)) {
+                        // Process exited while we were trying
+                        return false;
+                    }
+                }
+            } finally {
+                dte.SuppressUI = false;
+            }
+
+            // Another attempt, but display UI.
+            process.Attach();
+            return true;
+        }
+
+
+        private static string FixPath(string path) {
+            var dir = PathUtils.GetParent(path);
+            var newDir = GetFinalPathName(dir);
+            if (newDir != dir) {
+                // Minimize normalization that happens here
+                path = PathUtils.TrimEndSeparator(newDir) + "\\" + PathUtils.GetFileOrDirectoryName(path);
+            }
+
+            if (Path.IsPathRooted(path)) {
+                var root = Path.GetPathRoot(path);
+                var newRoot = "/mnt/" + root.ToLowerInvariant().Replace(":", "");
+                path = newRoot + path.Substring(root.Length);
+            }
+
+            return path.Replace('\\', '/');
+        }
+
+        private static string QuoteSingleArgument(string arg) {
+            return "'{0}'".FormatInvariant(arg.Replace("'", "'\\''"));
```

Comment:
Bash single quotes [don't allow any escaping whatsoever](https://www.gnu.org/software/bash/manual/html_node/Single-Quotes.html, not even of nested single quotes.

To allow for arbitrary strings here, you probably need to use [this](https://www.gnu.org/software/bash/manual/html_node/ANSI_002dC-Quoting.html).